File: 07/2.hs
import qualified Data.Map as M
import Data.List.Split (splitOn)

type Color = String
type Rule = (Color,Int)
type Line = (Color, (M.Map Color Int))
type RuleSet = M.Map Color Int
type Data = M.Map Color RuleSet

main = do
    input <- lines <$> readFile "input.txt"
    let m = M.fromList $ map parseLine input
    let n = contains m "shiny gold"
    putStrLn $ show $ (n - 1)

contains :: Data -> Color -> Int
contains xs c = 1 + (sum $ map (\k -> (rules M.! k) * (contains xs k)) $ M.keys rules) where
    rules = xs M.! c

parseLine :: String -> Line
parseLine xs =
    let c = unwords $ fst $ span (/="bags") $ words xs
        rulesStr = unwords $ tail $ snd $ span (/="contain") $ words xs
        rules = if ("no" `elem` (words rulesStr)) then [] else map parseRule $ splitOn ", " rulesStr
        in (c, M.fromList rules)

parseRule :: String -> Rule
parseRule xs =
    let n = read $ head $ words xs
        c = unwords $ tail $ init $ words xs
        in (c,n)
    

File: 07/2.js
const input = require("fs").readFileSync("input.txt").toString().split("\n")
const obj = input.reduce(parseRule, {})

function parseRule(obj, line) {
    if (/no other/.test(line)) return obj
    
    const re = /((?:\w+\s)+)bags contain ((?:\d+\s+(?:\w+\s)+bags?[,|.]\s?)*)/
    const [_, color, rules] = re.exec(line)
    return {
        ...obj,
        [color.trim()]: rules.split(", ").reduce((acc,rule) => {
            const [_, num, color] = /(\d+)\s+((?:\w+\s)+?)bag/.exec(rule)
            return { ...acc, [color.trim()]: num }
        }, {})
    }
}

function count(obj, key) {
    if (!(key in obj)) return 1
    if (!isNaN(obj[key])) return +obj[key]
    else return 1 + Object.keys(obj[key]).map(k => +obj[key][k] * count(obj, k)).reduce((a,b) => a+b, 0)
}

const result = count(obj, "shiny gold") - 1 /* itself */
console.log(result)
